### PR TITLE
Fix slow and frozen frames tracking

### LINKED ISSUE #2264: Slow and Frozen frames are reported wrongly for Activity transactions
### Description

At the moment we report **all** frozen/slow frames that ever happened since the app start for each transaction, meaning, that the number of them just increases for each new transaction. We should instead report metrics per activity and not the overall amount from the entire app run.

### PR DESCRIPTION
**Pull request by adinauer**
## :scroll: Description

Fixes the wrong reporting of frames by snapshotting slow and frozen frame values at the start and end, then calculating a delta.

## :bulb: Motivation and Context

Fixes #2264 

## :green_heart: How did you test it?
Sample App; Unit Tests

## :pencil: Checklist

- [x] I reviewed the submitted code
- [x] I added tests to verify the changes
- [ ] I updated the docs if needed
- [x] No breaking changes

## :crystal_ball: Next steps
Once merged, release `6.4.3` then merge `6.4.x` into `main`.

### CODE REVIEW

File: CHANGELOG.md [hunk 1]
```diff
@@ -1,5 +1,9 @@
 # Changelog
 
+## Unreleased
+
+- Fix slow and frozen frames tracking ([#2271](https://github.com/getsentry/sentry-java/pull/2271))
```
Comment (adinauer) [inline CHANGELOG.md:5]: @marandaneto since no public API changes were necessary I assume this fix doesn't require any further action on hybrid SDKs.

**Comment by romtsn** [comment]
> I just did a review for this PR and looked at the `ActivityLifecycleIntegration` to find out how this code works together in combination with it. We thought that we can't call `frameMetricsAggregator.reset()`, because `if users disable auto-finish of the activity transactions and finish them manually in onStop.`; see [#2269 (comment)](https://github.com/getsentry/sentry-java/pull/2269#discussion_r984804580)
> 
> If I'm not mistaken, we can't have multiple auto generated activity transactions running in parallel, because call `stopPreviousTransactions` in here
> 
> https://github.com/getsentry/sentry-java/blob/730095684bb732a0a7d0cc88d3c3b913a08d7cfe/sentry-android-core/src/main/java/io/sentry/android/core/ActivityLifecycleIntegration.java#L152-L156
> 
> So either we keep the code in `ActivityLifecycleIntegration` and we can safely call `frameMetricsAggregator.reset()` or we also need to adapt the code above.

The problem is if the transaction still has unfinished children, then calling `stopPreviousTransaction` won't actually finish them until the children are finished. But when the children get finished, we already reset the frame metrics, so the transaction won't have them at all.

**Comment by philipphofmann** [comment]
> The problem is if the transaction still has unfinished children, then calling stopPreviousTransaction won't actually finish them until the children are finished. But when the children get finished, we already reset the frame metrics, so the transaction won't have them at all.

Ah, I missed that one, thanks for pointing that out.

File: sentry-android-core/src/test/java/io/sentry/android/core/ActivityFramesTrackerTest.kt [hunk 2]
```diff
@@ -102,15 +110,166 @@ class ActivityFramesTrackerTest {
...
+    }
+
+    @Test
+    fun `different activities have separate counts - even when called out of order`() {
+        val sut = fixture.getSut()
+        val array0 = SparseIntArray().also {
+            it.put(1, 100)
+            it.put(20, 5)
+            it.put(705, 6)
+        }.let { arrayOf(it) }
```
Comment (philipphofmann) [inline sentry-android-core/src/test/java/io/sentry/android/core/ActivityFramesTrackerTest.kt:146]: `l`: A bit shorter
```suggestion
        val array0 = SparseIntArray().apply {
            put(1, 100)
            put(20, 5)
            put(705, 6)
        }.let { arrayOf(it) }
```

File: sentry-android-core/src/test/java/io/sentry/android/core/ActivityFramesTrackerTest.kt [hunk 3]
```diff
@@ -102,15 +110,166 @@ class ActivityFramesTrackerTest {
...
+        val array3 = SparseIntArray().also {
+            it.put(1, 135)
+            it.put(20, 11)
+            it.put(705, 12)
+        }.let { arrayOf(it) }
+
+        val otherActivity = mock<Activity>()
+        val otherSentryId = SentryId()
+
+        whenever(fixture.aggregator.metrics).thenReturn(array0, array1, array2, array3)
```
Comment (philipphofmann) [inline sentry-android-core/src/test/java/io/sentry/android/core/ActivityFramesTrackerTest.kt:166]: `l`: It would be easier to understand if these arrays are named something like `activity1AddFrameMetrics`, `activity1StopFrameMetrics`, ...

File: sentry-android-core/src/test/java/io/sentry/android/core/ActivityFramesTrackerTest.kt [hunk 4]
```diff
@@ -102,15 +110,166 @@ class ActivityFramesTrackerTest {
...
+        val sut = fixture.getSut()
+        val array0 = SparseIntArray().also {
+            it.put(1, 100)
+            it.put(20, 5)
+            it.put(705, 6)
+        }.let { arrayOf(it) }
+        val array1 = SparseIntArray().also {
+            it.put(1, 110)
+            it.put(20, 6)
+            it.put(705, 7)
```
Comment (philipphofmann) [inline sentry-android-core/src/test/java/io/sentry/android/core/ActivityFramesTrackerTest.kt:150]: `m`: Testing with boundary values of the equivalence classes `[0-16] [17-700] [701 - infinity]` makes the tests a bit better. This helps to detect off by 1 errors.
```suggestion
            it.put(16, 110)
            it.put(17, 3) // Splitted up 20 
            it.put(700, 3)
            it.put(701, 7)
```

File: sentry-android-core/src/test/java/io/sentry/android/core/ActivityFramesTrackerTest.kt [hunk 5]
```diff
@@ -102,15 +110,166 @@ class ActivityFramesTrackerTest {
...
+        sut.addActivity(fixture.activity)
+        sut.addActivity(otherActivity)
+        sut.setMetrics(fixture.activity, fixture.sentryId)
+        sut.setMetrics(otherActivity, otherSentryId)
+
+        val metrics = sut.takeMetrics(fixture.sentryId)
+        val otherMetrics = sut.takeMetrics(otherSentryId)
+
+        val totalFrames = metrics!!["frames_total"]
+        assertEquals(totalFrames!!.value, 21f) // 15 + 3 + 3
```
Comment (philipphofmann) [inline sentry-android-core/src/test/java/io/sentry/android/core/ActivityFramesTrackerTest.kt:177]: `l`: Instead of a comment here, we could declare variables before the arrays, use them when defining the number of frames in the arrays, and then sum them together here. This would also make changes to the tests easier.

File: sentry-android-core/src/test/java/io/sentry/android/core/ActivityFramesTrackerTest.kt [hunk 6]
```diff
@@ -102,15 +110,166 @@ class ActivityFramesTrackerTest {
         assertEquals(slowFrames!!.value, 5f)
     }
 
+    @Test
+    fun `sets slow and frozen frames even if start was null`() {
+        val sut = fixture.getSut()
+        val arrayAll = SparseIntArray()
+        arrayAll.put(1, 100)
+        arrayAll.put(20, 5)
+        arrayAll.put(705, 6)
```
Comment (philipphofmann) [inline sentry-android-core/src/test/java/io/sentry/android/core/ActivityFramesTrackerTest.kt:119]: `l`:
```suggestion
        val arrayAll = SparseIntArray().apply {
          put(1, 100)
          put(20, 5)
          put(705, 6)
        }
```

File: sentry-android-core/src/test/java/io/sentry/android/core/ActivityFramesTrackerTest.kt [hunk 7]
```diff
@@ -102,15 +110,166 @@ class ActivityFramesTrackerTest {
...
+
+        whenever(fixture.aggregator.metrics).thenReturn(array0, array1, array2, array3)
+
+        sut.addActivity(fixture.activity)
+        sut.addActivity(otherActivity)
+        sut.setMetrics(fixture.activity, fixture.sentryId)
+        sut.setMetrics(otherActivity, otherSentryId)
+
+        val metrics = sut.takeMetrics(fixture.sentryId)
+        val otherMetrics = sut.takeMetrics(otherSentryId)
```
Comment (philipphofmann) [inline sentry-android-core/src/test/java/io/sentry/android/core/ActivityFramesTrackerTest.kt:174]: `l`: If we do this, we don't need them everywhere below.
```suggestion
        val metrics = sut.takeMetrics(fixture.sentryId)!!
        val otherMetrics = sut.takeMetrics(otherSentryId)!!
```

File: sentry-android-core/src/test/java/io/sentry/android/core/ActivityFramesTrackerTest.kt [hunk 8]
```diff
@@ -150,6 +318,15 @@ class ActivityFramesTrackerTest {
         sut.stop()
     }
 
+    @Test
+    fun `stop resets frameMetricsAggregator`() {
+        val sut = fixture.getSut()
+
+        sut.stop()
+
+        verify(fixture.aggregator, times(1)).reset()
```
Comment (philipphofmann) [inline sentry-android-core/src/test/java/io/sentry/android/core/ActivityFramesTrackerTest.kt:327]: `l`: `times(1)` is the default.
```suggestion
        verify(fixture.aggregator).reset()
```

File: sentry-android-core/src/main/java/io/sentry/android/core/ActivityFramesTracker.java [hunk 9]
```diff
@@ -27,6 +28,8 @@ public final class ActivityFramesTracker {
 
   private final @NotNull Map<SentryId, Map<String, @NotNull MeasurementValue>>
       activityMeasurements = new ConcurrentHashMap<>();
+  private final @NotNull Map<Activity, FrameCounts> frameCountAtStartSnapshots =
```
Comment (philipphofmann) [inline sentry-android-core/src/main/java/io/sentry/android/core/ActivityFramesTracker.java:31]: I like the name 👏

File: sentry-android-core/src/main/java/io/sentry/android/core/ActivityFramesTracker.java [hunk 10]
```diff
@@ -55,33 +58,50 @@ public synchronized void addActivity(final @NotNull Activity activity) {
...
+      frameCountAtStartSnapshots.put(activity, frameCounts);
+    }
+  }
+
+  private @Nullable FrameCounts diffFrameCountsAtEnd(final @NotNull Activity activity) {
+    @Nullable
+    final FrameCounts frameCountsAtStartFromMap = frameCountAtStartSnapshots.remove(activity);
+    @NotNull
+    final FrameCounts frameCountsAtStart =
+        frameCountsAtStartFromMap == null ? new FrameCounts(0, 0, 0) : frameCountsAtStartFromMap;
```
Comment (philipphofmann) [inline sentry-android-core/src/main/java/io/sentry/android/core/ActivityFramesTracker.java:76]: `m`: When do we run into this edge case that the `frameCountsAtStartFromMap` is `null`? We already check that the activity is not `null` here https://github.com/getsentry/sentry-java/blob/f6b0dc7a17408741396774173c2637b717e3cbe1/sentry-android-core/src/main/java/io/sentry/android/core/ActivityLifecycleIntegration.java#L169-L172
The parameter is also annotated `@NotNull`. If this happens, something is wrong and I think we should better not count the frames at all and return `null` here. If we use `FrameCounts(0, 0, 0)`, I think we risk calculating incorrect frame counts.

File: sentry-android-core/src/main/java/io/sentry/android/core/ActivityFramesTracker.java [hunk 11]
```diff
@@ -55,33 +58,50 @@ public synchronized void addActivity(final @NotNull Activity activity) {
...
-  public synchronized void setMetrics(
-      final @NotNull Activity activity, final @NotNull SentryId sentryId) {
+  private void snapshotFrameCountsAtStart(final @NotNull Activity activity) {
+    FrameCounts frameCounts = calculateCurrentFrameCounts();
+    if (frameCounts != null) {
+      frameCountAtStartSnapshots.put(activity, frameCounts);
+    }
+  }
+
+  private @Nullable FrameCounts diffFrameCountsAtEnd(final @NotNull Activity activity) {
```
Comment (philipphofmann) [inline sentry-android-core/src/main/java/io/sentry/android/core/ActivityFramesTracker.java:71]: `l`: I would move this method below `setMetrics`, because it is the only function calling this method. Now when I read `setMetrics` I have to jump all the way up to here to find the method. Keeping functions calling each other close to each other in the code improves readability, IMO.

File: sentry-android-core/src/main/java/io/sentry/android/core/ActivityFramesTracker.java [hunk 12]
```diff
@@ -100,13 +120,53 @@ public synchronized void setMetrics(
...
+    private FrameCounts(final int totalFrames, final int slowFrames, final int frozenFrames) {
+      this.totalFrames = totalFrames;
+      this.slowFrames = slowFrames;
+      this.frozenFrames = frozenFrames;
+    }
+  }
+
+  @SuppressWarnings("NullAway")
+  public synchronized void setMetrics(
+      final @NotNull Activity activity, final @NotNull SentryId sentryId) {
```
Comment (philipphofmann) [inline sentry-android-core/src/main/java/io/sentry/android/core/ActivityFramesTracker.java:140]: `l`: I think this would improve readability a bit. I had to jump to here https://github.com/getsentry/sentry-java/blob/f6b0dc7a17408741396774173c2637b717e3cbe1/sentry-android-core/src/main/java/io/sentry/android/core/ActivityLifecycleIntegration.java#L172 to understand which id it is. Same applies for `takeMetrics`.
```suggestion
      final @NotNull Activity activity, final @NotNull SentryId transactionId) {
```

**Review by philipphofmann — COMMENTED** [review]
One edge case maybe to fix. Almost LGTM. Thanks for fixing this, @adinauer 👏 .

I added plenty of nitpicks; you can ignore them if you want. Just minor things to improve the readability of the code a bit. Some of them could also be personal preference.

File: sentry-android-core/src/test/java/io/sentry/android/core/ActivityFramesTrackerTest.kt (hunk 2 above)
Comment (adinauer) [reply sentry-android-core/src/test/java/io/sentry/android/core/ActivityFramesTrackerTest.kt:146]: Yeah I prefer `also` over `apply`. Fewer surprises this way. Gonna keep it this way if you don't mind :D

File: sentry-android-core/src/main/java/io/sentry/android/core/ActivityFramesTracker.java (hunk 10 above)
Comment (adinauer) [reply sentry-android-core/src/main/java/io/sentry/android/core/ActivityFramesTracker.java:76]: My idea here was that we might not be able to determine frame counts if the `FrameMetricsAggregator` was freshly created as `getMetrics()` on it is marked with `@Nullable`. Since I changed some code around it shouldn't be necessary as we have a `FrameCounts(0, 0, 0)`. Gonna change it so it doesn't calculate a difference if it can't find a start snapshot.

File: sentry-android-core/src/test/java/io/sentry/android/core/ActivityFramesTrackerTest.kt (hunk 2 above)
Comment (philipphofmann) [reply sentry-android-core/src/test/java/io/sentry/android/core/ActivityFramesTrackerTest.kt:146]: No strong opinion. It's `l`.

File: sentry-android-core/src/test/java/io/sentry/android/core/ActivityFramesTrackerTest.kt (hunk 4 above)
Comment (adinauer) [reply sentry-android-core/src/test/java/io/sentry/android/core/ActivityFramesTrackerTest.kt:150]: Just used the values that were there in old tests; can update

File: sentry-android-core/src/test/java/io/sentry/android/core/ActivityFramesTrackerTest.kt (hunk 3 above)
Comment (adinauer) [reply sentry-android-core/src/test/java/io/sentry/android/core/ActivityFramesTrackerTest.kt:166]: Gonna update

File: sentry-android-core/src/main/java/io/sentry/android/core/ActivityFramesTracker.java (hunk 11 above)
Comment (adinauer) [reply sentry-android-core/src/main/java/io/sentry/android/core/ActivityFramesTracker.java:71]: Thanks, forgot to reorder them

File: sentry-android-core/src/test/java/io/sentry/android/core/ActivityFramesTrackerTest.kt (hunk 5 above)
Comment (adinauer) [reply sentry-android-core/src/test/java/io/sentry/android/core/ActivityFramesTrackerTest.kt:177]: Hmm the test is already pretty long. Not sure if extracting and doing calculations improves readability over the comment or makes it even more complicated.

File: sentry-android-core/src/main/java/io/sentry/android/core/ActivityFramesTracker.java [hunk 13]
```diff
@@ -100,39 +101,98 @@ public synchronized void setMetrics(
...
     measurements.put("frames_total", tfValues);
     measurements.put("frames_slow", sfValues);
     measurements.put("frames_frozen", ffValues);
 
-    activityMeasurements.put(sentryId, measurements);
+    activityMeasurements.put(transactionId, measurements);
+  }
+
+  private @Nullable FrameCounts diffFrameCountsAtEnd(final @NotNull Activity activity) {
+    @Nullable final FrameCounts frameCountsAtStart = frameCountAtStartSnapshots.remove(activity);
```
Comment (philipphofmann) [inline sentry-android-core/src/main/java/io/sentry/android/core/ActivityFramesTracker.java:148]: Thanks 🙏

**Review by philipphofmann — APPROVED** [review]
Thank you @adinauer 👏

**Review by romtsn — APPROVED** [review]
LGTM 👍
